---
title: " Oracle soils Burkina Faso t = 0 (#5) "
author: "Frédéric Mahé"
date: '`r format(Sys.time(), "%d %B %Y")`'

output:
  rmarkdown::html_document:
    theme: lumen
    toc: yes
    toc_float: TRUE
    keep_md: yes
    # code_folding: hide
---

```{r setup, include=FALSE}
rm(list = ls()) # remove all the object before starting
knitr::opts_chunk$set(echo = TRUE)
```

***

#### load required packages

```{r packages, message=FALSE}
library(here)
library(tidyverse)
library(vegan)
```


# SSU rRNA metabarcoding (16S V3-V4)

## preamble

### variables and functions

```{r}
input <- "Oracle_soils_16S_341F_785R_87_samples.OTU.filtered.cleaved.nosubstringOTUs.mumu.table3"
title <- "Oracle MetaB t0 SSU NMDS (80 samples)"
useful_metadata <- c("OTU", "total", "amplicon", "length", "abundance",
                     "spread", "identity", "taxonomy", "references")
cbPalette <- c("#E69F00", "#56B4E9", "#009E73", "#0072B2", "#D55E00", "#CC79A7")
seed <- 123
data_folder <- "MetaB_t0"
```

```{r}
load_clean_16S_occurrence_data <- function(folder, filename){
    here::here("data", folder, filename) %>%
        read_tsv(show_col_types = FALSE)
}
```

### load data

```{r}
load_clean_16S_occurrence_data(folder = data_folder,
                               filename = input) -> full_occurrence_table
```


## NMDS

### computation

extract occurrence values, transpose for vegan, dissimilarity matrix
(Bray-Curtis), ordination:

```{r}
set.seed(seed)

full_occurrence_table %>%
    select(-all_of(useful_metadata)) %>%
    t() %>%
    vegan::vegdist(., method = "bray") %>%
    vegan::metaMDS() -> occurrences_t.bray.nmds
```


### prepare results

extract data scores and add stress value:

```{r}
occurrences_t.bray.nmds %>%
    vegan::scores(., display = "sites") %>%
    as.data.frame() %>%
    rownames_to_column(var = "samples") -> data.scores

x_min <- min(data.scores$NMDS1)
y_max <- max(data.scores$NMDS2)

stress <- occurrences_t.bray.nmds$stress
stress_annotation <- paste("stress: ", round(stress, digits = 4), sep = "")
```

parse sample names:

```{r}
data.scores %>%
    separate(col = samples, into = c("country", "locality", "sample"), sep = "-") -> d
```


### plot

```{r}
ggplot(data = d,
       aes(x = NMDS1, y = NMDS2, fill = locality)) +
    scale_fill_manual(values = cbPalette) +
    scale_colour_manual(values = cbPalette) +
    geom_point(size = 6, shape = 21, colour = "black", stroke = 0.5) +
    theme_bw(base_size = 16) +
    ggtitle(title) +
    annotate("text", x = x_min + abs(x_min / 9),
             y = y_max, label = stress_annotation) -> nmds_plot

nmds_plot
## ggsave(output, height = height, width = width)
```

No obvious structuration.

Add ellipses representing t-distributions (solid lines) and normal
distributions (dashed lines):

```{r}
nmds_plot +
    stat_ellipse(aes(colour = locality), type = "norm", linetype = 2) +
    stat_ellipse(aes(colour = locality), type = "t", linetype = 1)
```

The t-distribution, also known as Student's t-distribution, is a type
of normal distribution used for smaller sample sizes, where the
variance in the data is unknown.


Clean up:

```{r}
rm(data.scores, full_occurrence_table, input, nmds_plot,
   occurrences_t.bray.nmds, samples, stress, stress_annotation, title,
   useful_metadata, x_min, y_max)
```


# shotgun metagenomics (SSU rRNA assignment with kraken)

## preamble

### variables and functions

```{r}
data_folder <- "MetaG_t0"
input <- "oracle_metaGt0_79_samples_bracken_SSU_assignment_genuses_rarefied.table3"
title <- "Oracle MetaG t0 SSU NMDS (79 samples)"
useful_metadata <- c("taxonomy")
```


### load data

```{r}
load_clean_16S_occurrence_data(folder = data_folder,
                               filename = input) -> full_occurrence_table
```


## NMDS

### computation

extract occurrence values, transpose for vegan, dissimilarity matrix
(Bray-Curtis), ordination:

```{r}
set.seed(seed)

full_occurrence_table %>%
    select(-all_of(useful_metadata)) %>%
    t() %>%
    vegan::vegdist(., method = "bray") %>%
    vegan::metaMDS() -> occurrences_t.bray.nmds
```


### prepare results

extract data scores and add stress value:

```{r}
occurrences_t.bray.nmds %>%
    vegan::scores(., display = "sites") %>%
    as.data.frame() %>%
    rownames_to_column(var = "samples") -> data.scores

x_min <- min(data.scores$NMDS1)
y_max <- max(data.scores$NMDS2)

stress <- occurrences_t.bray.nmds$stress
stress_annotation <- paste("stress: ", round(stress, digits = 4), sep = "")
```

parse sample names:

```{r}
data.scores %>%
    separate(col = samples, into = c("country", "locality", "sample"), sep = "-") -> d
```


### plot

```{r}
ggplot(data = d,
       aes(x = NMDS1, y = NMDS2, fill = locality)) +
    scale_fill_manual(values = cbPalette) +
    scale_colour_manual(values = cbPalette) +
    geom_point(size = 6, shape = 21, colour = "black", stroke = 0.5) +
    theme_bw(base_size = 16) +
    ggtitle(title) +
    annotate("text", x = x_min + abs(x_min / 9),
             y = y_max, label = stress_annotation) -> nmds_plot

nmds_plot
```

There seems to be a very strong batch effect between sequencing runs
SAMA12 and SAMA21.

Add ellipses representing t-distributions (solid lines) and normal
distributions (dashed lines):

```{r}
nmds_plot +
    stat_ellipse(aes(colour = locality), type = "norm", linetype = 2) +
    stat_ellipse(aes(colour = locality), type = "t", linetype = 1)
```

The t-distribution, also known as Student's t-distribution, is a type
of normal distribution used for smaller sample sizes, where the
variance in the data is unknown.


***

```{r}
sessionInfo()
rm(list = ls())
```
