---
title: " Oracle soils Burkina Faso t = 0 (#7) "
author: "Frédéric Mahé"
date: '`r format(Sys.time(), "%d %B %Y")`'

output:
  rmarkdown::html_document:
    theme: lumen
    toc: yes
    toc_float: TRUE
    keep_md: yes
    # code_folding: hide
---

```{r setup, include=FALSE}
rm(list = ls()) # remove all the object before starting
knitr::opts_chunk$set(echo = TRUE)
```

***

#### load required packages

```{r packages, message=FALSE}
library(here)
library(tidyverse)
```


# SSU rRNA metabarcoding (16S V3-V4)

## preamble

### download Silva files

create folders if need be:

```{r}
dir.create(here::here("data", "Silva"),
           recursive = TRUE,
           showWarnings = FALSE)
```

if the Silva files are not already present, download them:

```{r}
download_if_missing <- function(file_name, url) {
    local_file <- here::here("data", "Silva", file_name)
    if (! file.exists(local_file)) {
        download.file(url = paste0(url, file_name),
                      destfile = local_file,
                      method = "auto")
    }
}

URL <- "https://www.arb-silva.de/fileadmin/silva_databases/release_138.1/Exports/"

download_if_missing(file_name = "SILVA_138.1_SSURef_NR99_tax_silva.fasta.gz",
                    url = URL)
download_if_missing(file_name = "tax_slv_ssu_138.1.txt.gz",
                    url = paste0(URL, "taxonomy/"))
```

clean up:

```{r}
rm(download_if_missing)
```


### Silva ranked taxonomic data

#### variables and functions

```{r}
column_names <- c("taxonomic_paths", "taxid", "taxonomic_rank", "remark", "release")
columns_to_keep <- c("taxonomic_paths", "taxonomic_rank")
```

```{r}
load_tax_slv_data <- function(tax_slv_data) {
    here::here("data", "Silva", tax_slv_data) %>%
        read_tsv(col_names = column_names,
                 col_types = "cifcf",
                 col_select = all_of(columns_to_keep),
                 show_col_types = FALSE)
}

. %>%
    mutate(taxonomic_paths = str_remove(taxonomic_paths, ";$"),
           taxonomic_paths = str_remove(taxonomic_paths, ".*;")) %>%
    arrange(taxonomic_paths) %>%
    count(taxonomic_paths, taxonomic_rank) -> link_taxa_and_ranks

. %>%
    count(taxonomic_paths) %>%
    filter(n > 1) -> find_duplicated_ranks

. %>%
    filter(! taxonomic_paths %in% c("Incertae Sedis", "uncultured") &
           ! (taxonomic_paths == "SAR" & taxonomic_rank == "phylum") &
           ! (taxonomic_paths == "Stramenopiles" & taxonomic_rank == "subphylum") &
           ! (taxonomic_paths == "Labyrinthulomycetes" & taxonomic_rank == "class")) %>%
    select(-n) -> ban_some_taxa_and_ranks
```

#### load data

```{r}
load_tax_slv_data("tax_slv_ssu_138.1.txt.gz") %>%
    link_taxa_and_ranks -> d
```

#### find and eliminate conflicts

A given taxa name should be present in only one of the 20 possible
ranks. Find taxa names that do not follow that rule:

```{r}
d %>%
    find_duplicated_ranks
```

Besides 'uncultured' and 'Incertae Sedis', all conflicts come from
these four entries:

- Eukaryota;Amorphea;Amoebozoa;SAR;	46930	phylum		138
- Eukaryota;Amorphea;Amoebozoa;SAR;Stramenopiles;	46931	subphylum		138
- Eukaryota;Amorphea;Amoebozoa;SAR;Stramenopiles;Labyrinthulomycetes;	46932	class		138
- Eukaryota;Amorphea;Amoebozoa;SAR;Stramenopiles;Labyrinthulomycetes;Sorodiplophrys;	46933	genus		138

handle these specific cases:

```{r}
d %>%
    ban_some_taxa_and_ranks -> taxa2ranks
```

#### check final 'taxa to rank' table

how many taxonomic levels in Silva? 20 levels expected

```{r}
taxa2ranks %>%
    select(taxonomic_rank) %>%
    distinct()
```

clean up:

```{r}
rm(find_duplicated_ranks, d, ban_some_taxa_and_ranks,
   columns_to_keep, link_taxa_and_ranks, load_tax_slv_data,
   URL, column_names)
```


### Silva SSU NR99 references

parse reference fasta sequences, keep only headers. Note that the
accession number and taxonomic paths are separated by a space, but
additional spaces can appear in taxonomic paths. That's why I replace
the first space with '@', and then use '@' as a field separator.

```{r, warning = FALSE}
silva_SSURef_NR99 <- "SILVA_138.1_SSURef_NR99_tax_silva.fasta.gz"
preprocessed_silva_SSURef_NR99 <- "SILVA_138.1_SSURef_NR99_tax_silva.txt.gz"

if (file.exists(here::here("data", "Silva", preprocessed_silva_SSURef_NR99))) {
    here::here("data", "Silva", preprocessed_silva_SSURef_NR99) %>%
        read_tsv(col_names = c("references", "taxonomic_paths"),
                 show_col_types = FALSE) -> references2taxonomic_paths
} else {
    ## very slow and potentially broken, prefer shell code (see README.md)
    here::here("data", "Silva", silva_SSURef_NR99) %>%
        read_delim(delim = ">",
                   col_names = c("sequence", "header"),
                   col_select = c(header), show_col_types = FALSE) %>%
        drop_na() %>%
        mutate(header = str_replace(header, " ", "@")) %>%
        separate(col = header,
                 into = c("references", "taxonomic_paths"),
                 sep = "@") -> references2taxonomic_paths
}

rm(silva_SSURef_NR99, preprocessed_silva_SSURef_NR99)
```


### 16S occurrence data

#### variables and functions

```{r}
input <- "Oracle_soils_16S_341F_785R_87_samples.OTU.filtered.cleaved.nosubstringOTUs.mumu.table3"
new_taxonomy_table <- str_replace(input, "table3", "table4")
metadata_file <- "metadata.txt"
useful_metadata <- c("OTU", "total", "amplicon", "length", "abundance",
                     "spread", "identity", "taxonomy", "references")
seed <- 123
```

```{r}
load_clean_16S_occurrence_data <- function(filename){
    here::here("data", "MetaB_t0", filename) %>%
        read_tsv(na = c("", "NA", "No_hit", "0", "0.0"),
                 n_max = 10000,
                 show_col_types = FALSE)
}

load_metadata <- function(filename){
    ## replace '_' and remove trailing zeros in soil codes
    here::here("data", filename) %>%
        read_tsv(show_col_types = FALSE) %>%
        mutate(soil_code = str_replace_all(soil_code, "_", "-"),
               soil_code = str_replace(soil_code, "-0(?=[1-9]$)", "-"))
}

. %>%
    select(OTU, identity, references, starts_with("BU-")) %>%
    pivot_longer(cols = starts_with("BU-"),
                 names_to = "soil_code",
                 values_to = "reads",
                 values_drop_na = TRUE) %>%
    separate_rows(references, sep = ",") %>%
    group_by(OTU, soil_code) %>%
    mutate(weight = 1 / n()) %>%
    ungroup() %>%
    mutate(reads = reads * weight) %>%
    select(-weight) -> compute_weight_per_reference

. %>%
    left_join(references2taxonomic_paths,
              by = "references") %>%
    separate_rows(taxonomic_paths, sep = ";") -> add_taxonomic_data

. %>%
    left_join(taxa2ranks, by = "taxonomic_paths") %>%
    filter(! is.na(taxonomic_rank)) %>%
    distinct() -> add_taxonomic_ranks

. %>%
    pivot_wider(names_from = taxonomic_rank,
                values_from = taxonomic_paths) %>%
    select(-OTU, -identity, -references) %>%
    count(soil_code, domain, phylum, class, order, family, genus,
          wt = reads, name = "reads") -> group_taxa_per_sample

. %>%
    left_join(metadata, by = "soil_code") -> add_metadata
```


#### load occurrence data

```{r}
load_clean_16S_occurrence_data(input) -> full_occurrence_table
load_metadata(metadata_file) -> metadata
```


## taxonomy

### weight taxonomic assignments

The goal is to avoid stars '*', by going back to the situation before
the strict assignment to the last common ancestor.

For example, let's assume that `OTU1` appears in `Sample1` with 9
reads, and that `OTU1` is equidistant to three reference sequences
with the following taxonomic assignments:

- reference1: genusA, speciesX
- reference2: genusA, speciesY
- reference3: genusB, speciesZ

`OTU1` is split and each reference receives 3 reads (9 divided by the
number of matching references). At that point, the OTU information is
dropped. Taxonomic assignments are then grouped by genus (and higher
ranks; species ranks is dropped). In the final occurrence table,
`GenusA` is present in `Sample1` with 6 reads, and `GenusB` with 3
reads.

Note: there are cases of a duplicated level in some Silva entries

```
>X92360.1.1415 Bacteria;Actinobacteriota;Actinobacteria;Frankiales;Geodermatophilaceae;Geodermatophilus;Geodermatophilus
```

Here the genus name is repeated. A call to `distinct()` solves the
issue, but that's not ideal.

Note: slv is limited to genus, no species rank.

```{r}
full_occurrence_table %>%
    compute_weight_per_reference %>%
    add_taxonomic_data %>%
    add_taxonomic_ranks %>%
    group_taxa_per_sample %>%
    add_metadata -> taxa_per_sample_table
```

```{r}
taxa_per_sample_table

taxa_per_sample_table %>%
    write_tsv(file = here::here("data", "MetaB_t0", new_taxonomy_table))

rm(taxa2ranks, references2taxonomic_paths)
ls()
```

### sanity check

The number of reads before and after the 'weight taxonomic
assignments' operation should be the same (maybe with a small margin
due to rounding errors):

```{r}
full_occurrence_table %>%
    select(OTU, starts_with("BU-")) %>%
    pivot_longer(cols = starts_with("BU-"),
                 names_to = "soil_code",
                 values_to = "reads",
                 values_drop_na = TRUE) %>%
    summarize(sum = sum(reads)) %>%
    pull() -> number_reads_before

taxa_per_sample_table %>%
    summarize(sum = sum(reads)) %>%
    pull() -> number_reads_after

abs(number_reads_before - number_reads_after) -> reads_variation

rm(number_reads_before, number_reads_after)
```

As expected, the weight taxonomic assignments operation has no
significant effect on the number of reads (variation is `r
reads_variation`).


roadmap:

- NA cells should be marked as unknown?,
- save final taxonomic table,
- add taxonomic barcharts,
- save table to results,
- move nitrogen stuff to a new Rmd file,
- add '(Frankia)' to the legend 'Actinorhizal N-Fixer', as there is
  only one genus with that function,
- remove n_max limitations,

***

```{r}
sessionInfo()
rm(list = ls())
```
